File: src/lib60870.netcore/iec104client/TagsCreation.cs
/*
 * IEC 60870-5-104 Client Protocol driver for {json:scada}
 * {json:scada} - Copyright (c) 2020 - 2026 - Ricardo L. Olsen
 * This file is part of the JSON-SCADA distribution (https://github.com/riclolsen/json-scada).
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 3.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

using System;
using MongoDB.Bson;
using MongoDB.Driver;
using lib60870.CS101;

namespace Iec10XDriver
{
    partial class MainClass
    {
        public static double AutoKeyMultiplier = 1000000;

        // Normalize time-tagged and short TypeIDs to their canonical base types.
        public static int MapAsduToBaseType(TypeID asdu)
        {
            switch ((int)asdu)
            {
                case 2:  case 30: return 1;   // Single Point (with/without timestamp)
                case 4:  case 31: return 3;   // Double Point
                case 6:  case 32: return 5;   // Step Position
                case 8:  case 33: return 7;   // Bitstring 32
                case 10: case 21: case 34: return 9;  // Measured Normalized
                case 12: case 35: return 11;  // Measured Scaled
                case 14: case 36: return 13;  // Measured Float
                case 16: case 37: return 15;  // Integrated Totals
                default: return (int)asdu;
            }
        }

        public static string Iec10xTypeDescription(int baseType)
        {
            switch (baseType)
            {
                case 1:  return "Single Point";
                case 3:  return "Double Point";
                case 5:  return "Step Position";
                case 7:  return "Bitstring 32";
                case 9:  return "Measured Normalized";
                case 11: return "Measured Scaled";
                case 13: return "Measured Float";
                case 15: return "Integrated Totals";
                case 20: return "Packed Single Point";
                default: return "ASDU " + baseType;
            }
        }

        // Allocates the next _id in the connection's dedicated partition [connNumber*1e6, (connNumber+1)*1e6).
        // First call queries MongoDB for the current max; subsequent calls just increment the cached value.
        public static double GetNextAutoKey(IEC10X_connection srv, IMongoCollection<BsonDocument> colRt)
        {
            if (srv.LastNewKeyCreated == 0)
            {
                double baseKey = srv.protocolConnectionNumber * AutoKeyMultiplier;
                double topKey  = (srv.protocolConnectionNumber + 1) * AutoKeyMultiplier;
                var filter = new BsonDocument("_id", new BsonDocument
                {
                    { "$gt", baseKey },
                    { "$lt", topKey }
                });
                var res = colRt
                    .Find(filter)
                    .Sort(Builders<BsonDocument>.Sort.Descending("_id"))
                    .Limit(1)
                    .FirstOrDefault();
                srv.LastNewKeyCreated = res != null ? res["_id"].ToDouble() + 1 : baseKey;
            }
            else
                srv.LastNewKeyCreated += 1;
            return srv.LastNewKeyCreated;
        }

        // Builds a complete realtimeData document for a new auto-created tag.
        // IEC clients never create command twins — monitor direction has no output-status TypeIDs.
        public static BsonDocument NewRealtimeTagDoc(IEC_Value iv, string connName, double id)
        {
            int    baseType   = MapAsduToBaseType(iv.asdu);
            string desc       = Iec10xTypeDescription(baseType);
            string tag        = connName + ";" + iv.common_address + ";" + iv.address;
            bool   isDigital  = iv.isDigital;
            return new BsonDocument
            {
                { "_id",         id },
                { "tag",         tag },
                { "type",        isDigital ? "digital" : "analog" },
                { "origin",      "supervised" },
                { "description", connName + "~" + desc + "~" + iv.address },
                { "ungroupedDescription", desc + " " + iv.address },
                { "group1",      connName },
                { "group2",      "CA " + iv.common_address },
                { "group3",      "" },
                { "protocolSourceConnectionNumber", (double)iv.conn_number },
                { "protocolSourceCommonAddress",    (double)iv.common_address },
                { "protocolSourceObjectAddress",    (double)iv.address },
                { "protocolSourceASDU",             (double)baseType },
                { "protocolSourceCommandDuration",  0.0 },
                { "protocolSourceCommandUseSBO",    false },
                { "commandOfSupervised",            0.0 },
                { "supervisedOfCommand",            0.0 },
                { "kconv1",  1.0 },
                { "kconv2",  0.0 },
                { "alarmState",    isDigital ? 2.0 : -1.0 },
                { "stateTextFalse", isDigital ? "FALSE" : "" },
                { "stateTextTrue",  isDigital ? "TRUE"  : "" },
                { "eventTextFalse", isDigital ? "FALSE" : "" },
                { "eventTextTrue",  isDigital ? "TRUE"  : "" },
                { "value",         iv.value },
                { "valueString",   iv.value.ToString() },
                { "invalid",       true },
                { "invalidDetectTimeout", 60000.0 },
                { "isEvent",       false },
                { "alarmDisabled", false },
                { "alarmed",       false },
                { "alerted",       false },
                { "alertState",    "" },
                { "annotation",    "" },
                { "commandBlocked", false },
                { "commissioningRemarks", "" },
                { "formula",       0.0 },
                { "frozen",        false },
                { "frozenDetectTimeout", 0.0 },
                { "hiLimit",       double.MaxValue },
                { "hihiLimit",     double.MaxValue },
                { "hihihiLimit",   double.MaxValue },
                { "loLimit",      -double.MaxValue },
                { "loloLimit",    -double.MaxValue },
                { "lololoLimit",  -double.MaxValue },
                { "historianDeadBand", 0.0 },
                { "historianPeriod",   0.0 },
                { "hysteresis",        0.0 },
                { "location",      BsonNull.Value },
                { "notes",         "" },
                { "overflow",      false },
                { "parcels",       BsonNull.Value },
                { "priority",      0.0 },
                { "protocolDestinations", new BsonArray() },
                { "sourceDataUpdate",  BsonNull.Value },
                { "substituted",       false },
                { "timeTag",           BsonNull.Value },
                { "timeTagAlarm",      BsonNull.Value },
                { "timeTagAtSource",   BsonNull.Value },
                { "timeTagAtSourceOk", false },
                { "transient",         false },
                { "unit",              "" },
                { "updatesCnt",        0.0 },
                { "valueDefault",      0.0 },
                { "zeroDeadband",      0.0 }
            };
        }
    }
}

File: src/OPC-UA-Client-Go/tags_creation.go
/*
 * OPC-UA Client Protocol driver for {json:scada}, in Go.
 * {json:scada} - Copyright (c) 2020-2026 - Ricardo L. Olsen
 * This file is part of the JSON-SCADA distribution (https://github.com/riclolsen/json-scada).
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 3.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

// Automatically created tag documents. Port of TagsCreation.cs.
//
// Every field of realtimeData is written out, at the same default the C#
// rtData class declares: the AdminUI and cs_data_processor both expect
// complete documents.

package main

import (
	"maps"
	"strconv"
	"strings"

	"github.com/riclolsen/json-scada/src/go-common/jstags"
	"go.mongodb.org/mongo-driver/v2/bson"
)

// tagTypeFor maps an OPC UA type name onto the json-scada tag type.
func tagTypeFor(asdu string, isArray bool) string {
	t := "analog"
	switch strings.ToLower(asdu) {
	case "boolean":
		t = "digital"
	case "string", "bytestring", "localeid", "localizedtext",
		"xmlelement", "qualifiedname", "guid":
		t = "string"
	case "nodeid", "expandednodeid", "extensionobject":
		t = "json"
	}
	if isArray {
		t = "json"
	}
	return t
}

// tagFromOPCParameters is the tag name of a supervised point.
func tagFromOPCParameters(ov OPCValue) string {
	return ov.ConnName + ";" + ov.Address
}

// newRealtimeDoc builds the realtimeData document for a discovered point.
//
// The C# version is five near-identical object literals; the differences
// between them are small enough to express as a base document plus the
// per-shape overrides below. Compare against TagsCreation.cs when changing
// anything here.
func newRealtimeDoc(ov OPCValue, id float64, commandOfSupervised float64) bson.M {
	typ := tagTypeFor(ov.Asdu, ov.IsArray)
	isCommand := ov.CreateCommandForMethod || ov.CreateCommandForSupervised

	doc := jstags.BaseDoc()
	maps.Copy(doc, bson.M{
		"_id":                              id,
		"protocolSourceBrowsePath":         ov.Path,
		"protocolSourceAccessLevel":        strconv.Itoa(int(ov.AccessLevels)),
		"protocolSourceASDU":               ov.Asdu,
		"protocolSourceCommonAddress":      ov.CommonAddress,
		"protocolSourceConnectionNumber":   float64(ov.ConnNumber),
		"protocolSourceObjectAddress":      ov.Address,
		"protocolSourceCommandUseSBO":      false,
		"protocolSourceCommandDuration":    0.0,
		"protocolSourcePublishingInterval": 5.0,
		"protocolSourceSamplingInterval":   2.0,
		"protocolSourceQueueSize":          10.0,
		"protocolSourceDiscardOldest":      true,

		"alarmState":           -1.0,
		"description":          ov.ConnName + "~" + ov.Path + "~" + ov.DisplayName,
		"ungroupedDescription": ov.DisplayName,
		"group1":               ov.ConnName,
		"group2":               ov.Path,
		"group3":               "",
		"stateTextFalse":       "",
		"stateTextTrue":        "",
		"eventTextFalse":       "",
		"eventTextTrue":        "",
		"origin":               "supervised",
		"tag":                  tagFromOPCParameters(ov),
		"type":                 typ,
		"value":                0.0,
		"valueString":          "",
		"valueJson":            "",

		"commandOfSupervised":  commandOfSupervised,
		"invalid":              true,
		"invalidDetectTimeout": 60000.0,
		"protocolDestinations": bson.A{},
		"supervisedOfCommand":  0.0,
	})

	switch {
	case isCommand:
		doc["origin"] = "command"
		doc["tag"] = tagFromOPCParameters(ov) + ";cmd"
		doc["description"] = ov.ConnName + "~" + ov.Path + "~" + ov.DisplayName + "-Command"
		doc["alarmState"] = 2.0
		doc["protocolSourcePublishingInterval"] = 0.0
		doc["protocolSourceSamplingInterval"] = 0.0
		doc["protocolSourceQueueSize"] = 0.0
		// parity: a command document carries no link back to the point it
		// drives unless it was created for a writable variable, and then
		// the supervised twin is the very next key.
		doc["commandOfSupervised"] = 0.0
		if ov.CreateCommandForSupervised {
			doc["supervisedOfCommand"] = id + 1
		}
		if typ == "digital" {
			doc["stateTextFalse"] = "FALSE"
			doc["stateTextTrue"] = "TRUE"
			doc["eventTextFalse"] = "FALSE"
			doc["eventTextTrue"] = "TRUE"
		}

	case typ == "digital":
		doc["alarmState"] = 2.0
		doc["stateTextFalse"] = "FALSE"
		doc["stateTextTrue"] = "TRUE"
		doc["eventTextFalse"] = "FALSE"
		doc["eventTextTrue"] = "TRUE"
		doc["value"] = ov.Value

	case typ == "string":
		doc["valueString"] = ov.ValueString

	case typ == "json":
		doc["valueString"] = ov.ValueString
		doc["valueJson"] = ov.ValueJSON

	default: // analog
		doc["value"] = ov.Value
		// parity: the analog literal in TagsCreation.cs is the only one
		// that never assigns valueJson, so the C# field keeps its null
		// default and the stored document gets null rather than "".
		doc["valueJson"] = nil
	}

	if isCommand {
		doc["value"] = 0.0
		doc["valueString"] = ""
		doc["valueJson"] = ""
	}

	return doc
}
